File: Cpp/Binary Representation.cpp
#include <string>
#include <stack>

class Solution {
public:
    /**
     *@param n: Given a decimal number that is passed in as a string
     *@return: A string
     */
    string intTobin(long long res) {
        if (res == 0) return "0";
        stack<char> stk;
        while (res) {
            stk.push(res % 2 + '0');
            res = res / 2;
        }
        string str;
        while (!stk.empty()) {
            str += stk.top();
            stk.pop();
        }
        return str;
    }
    string doublestr(string str) {
        int d = 0;
        string st = str;
        for (int i = str.size() - 1; i >= 0; i --) {
            int temp = (str[i] - '0') * 2 + d;
            st[i] = temp % 10 + '0';
            d = temp / 10;
        }
        return st;
    }
    string solve(string str) {
        int index = str.size() - 1;
        while (index >= 0 && str[index] == '0') {
            index --;
        }
        if (index < 0) return "";
        return str.substr(0, index + 1);
    }
    string binaryRepresentation(string n) {
        // wirte your code here
        int pos = -1;
        long long res = 0;
        for (int i = 0; i < n.size(); i ++) {
            if (n[i] != '.') {
                res = res * 10 + n[i] - '0';
            } else {
                pos = i;
                break;
            }
        }
        string str = intTobin(res);
        if (pos == -1) return str;
        string str1 = n.substr(pos + 1, n.size() - pos - 1), str2 = str;
        if (str1.back() != '0' && str1.back() != '5') return "ERROR";
        else str2 += ".";
        if (solve(str1) == "") return str;
        int time = 0, maxt = str1.size() * 2;
        while (str1.size() > 0) {
            if (time >= maxt) return "ERROR";
            if (str1[0] >= '5') str2 += "1";
            else str2 += "0";
            str1 = solve(doublestr(str1));
            time ++;
        }
        return str2;
    }
};

File: Java/Binary-Representation.java
public class Solution {
    /**
     *@param n: Given a decimal number that is passed in as a string
     *@return: A string
     */
    public String IntToBinary(int cur) {
        StringBuffer sb = new StringBuffer();
        if (cur == 0) return "0";
        while (cur != 0) {
            int temp = cur % 2;
            sb.append(String.valueOf(temp));
            cur = cur / 2;
        }
        sb.reverse();
        return sb.toString();
    }
    boolean isValid(String s) {
        if (s.charAt(s.length() - 1) != '5' && s.charAt(s.length() - 1) != '0') return false;
        return true;
    }
    boolean isZero(String s) {
        for (int i = 0; i < s.length(); i ++) {
            if (s.charAt(i) != '0') return false;
        }
        return true;
    }
    public String shift(String s) {
        int d = 0;
        StringBuffer sb = new StringBuffer();
        for (int i = s.length() -1; i >= 0; i --) {
            int temp = s.charAt(i) - '0';
            sb.append(String.valueOf((temp * 2 + d) % 10));
            d = (temp * 2 + d) / 10;
        }
        //while (sb.charAt(index) == '0') index ++;
        s = sb.reverse().toString();
        int index = sb.length() - 1;
        while (index >= 0 && s.charAt(index) == '0') index --;
        s = s.substring(0, index + 1);
        return s;
    }
    public String binaryRepresentation(String n) {
        // write your code here
        //考虑负数，考虑零，考虑整数。
        if (n == null || n.length() == 0) return "";
        int sign = 1;
        if (n.charAt(0) == '-') {
            n = n.substring(1);
            sign = -1;
        }
        int Point_index = -1;
        for (int i = 0; i < n.length(); i ++) {
            if (n.charAt(i) == '.') {
                Point_index = i;
                break;
            }
        }
        String res;
        if (Point_index == -1) {
            int cur = Integer.parseInt(n);
            res = IntToBinary(cur);
        } else {
            String Int_part = n.substring(0, Point_index);
            int Int_p = Integer.parseInt(Int_part);
            Int_part = IntToBinary(Int_p);
            String dem_part = n.substring(Point_index + 1);
            StringBuffer cur = new StringBuffer();
            int count = 0;
            while (count < 32 && dem_part.length() > 0) {
                if (!isValid(dem_part)) return "ERROR";
                if (isZero(dem_part)) break;
                if (dem_part.charAt(0) >= '5') cur.append("1");
                else cur.append("0");
                String s = shift(dem_part);
                dem_part = s;
                count ++;
            }
            if (!isZero(dem_part)) return "ERROR";
            if (cur.length() > 0) res = Int_part + "." + cur.toString();
            else res = Int_part;
        }
        if (sign == -1) return "-" + res;
        return res;
    }
}